### PR TITLE
tests: add new kernel objects tests

### PR DESCRIPTION
**Pull request by maksimmasalski**
Found out that important requirements are not tested by current
kernel objects tests. Decided to fix that situation

New added tests:
1. test_kobj_assign_perms_on_alloc_obj()
Create kernel object semaphore, dynamically allocate it from the
calling thread's resource pool.
Check that object's address is in bounds of that memory pool.
Then check the requestor thread will implicitly be assigned
permission on the allocated object by using
semaphore API k_sem_init()

2. test_no_ref_dyn_kobj_release_mem()
Dynamically allocated kernel objects whose access is controlled by
the permission system will use object permission as a reference count
If no threads have access to an object, the object's memory released.

3. test_krnl_obj_static_alloc_build_time()
Take addresses of the kernel objects which are statically allocated
during the build time and verify that they are not null.
That kernel objects shouldn't require manual
registration by the end user.

4. Clean-up. Removed unused variable from userspace test.

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>

### CODE REVIEW

**Comment by maksimmasalski** [comment]
I can't understand why Gitlint gives that error. I checked another tests use the same approach and define new typedefs.

**Comment by maksimmasalski** [comment]
@LeiW000 @andrewboie Could you review it? Do you know another variant, how to implement that test and avoid adding new structure?

**Comment by andrewboie** [comment]
> I can't understand why Gitlint gives that error. I checked another tests use the same approach and define new typedefs.

That other test is wrong.

If the only reason you are defining a typedef is to save yourself from typing the keyword `struct` then don't do it.

Normally we only define typedefs for types that have to be truly opaque or are part of the assembly offset generation mechanism.

**Review by tyrantlizard77 — DISMISSED** [review]
I started reviewing.

But you seem to be really confused on the memory address contained within a pointer variable, and the address of a pointer variable itself. This code is pervasively checking `&some_pointer` instead of its contents.

Please fix this.

File: tests/kernel/mem_protect/mem_protect/src/kobject.c [hunk 1]
```diff
@@ -1000,3 +1000,21 @@ void test_mark_thread_exit_uninitialized(void)
...
+ *
+ * @details Take addresses of the kernel objects which are statically allocated
+ * during the build time and verify that they are not null.
+ * That kernel objects shouldn't require manual registration by the end user.
+ *
+ * @ingroup kernel_memprotect_tests
+ */
+void test_krnl_obj_static_alloc_build_time(void)
+{
+	zassert_true(&child_stack != NULL, NULL);
```
[the remark this answers is not in the document]
Comment (maksimmasalski) [reply tests/kernel/mem_protect/mem_protect/src/kobject.c:1015]: @andrewboie Now I understand that I shouldn't use `&` operator at all. To review that kernel object is not NULL is enough to compare object itself with NULL value like `zassert_true(child_stack != NULL, NULL);`

[the remark this answers is not in the document]
Comment (andrewboie) [reply tests/kernel/mem_protect/mem_protect/src/kobject.c:1015]: stacks are arrays so it will NEVER be NULL.

[the remark this answers is not in the document]
Comment (maksimmasalski) [reply tests/kernel/mem_protect/mem_protect/src/kobject.c:1015]: For example to check kernel object semaphore I can access each field of struct and verify that it is not NULL
```
	struct k_sem *p = &kobject_sem;
	zassert_true(p->count != NULL, NULL);
```
I think the same I can do with stack

[the remark this answers is not in the document]
Comment (maksimmasalski) [reply tests/kernel/mem_protect/mem_protect/src/kobject.c:1015]: Wait, I think I can just check that pointer `zassert_true(p != NULL, NULL);`

[the remark this answers is not in the document]
Comment (maksimmasalski) [reply tests/kernel/mem_protect/mem_protect/src/kobject.c:1015]: @andrewboie That test I made to verify that kernel objects may be statically allocated at build time, without requiring any manual registration by the end user.

File: tests/kernel/mem_protect/obj_validation/src/main.c [hunk 2]
```diff
@@ -112,10 +121,109 @@ void test_generic_object(void)
...
+ *   the allocated object by using semaphore API k_sem_init()
+ *
+ * @ingroup kernel_memprotect_tests
+ *
+ * @see k_object_alloc()
+ */
+void test_kobj_assign_perms_on_alloc_obj(void)
+{
+	test_dyn_sem0 = k_object_alloc(K_OBJ_SEM);
+	zassert_not_null(&test_dyn_sem0, "Cannot allocate sem k_object");
```
[the remark this answers is not in the document]
Comment (maksimmasalski) [reply tests/kernel/mem_protect/obj_validation/src/main.c:142]: @andrewboie I emailed you about that test

[the remark this answers is not in the document]
Comment (andrewboie) [reply tests/kernel/mem_protect/obj_validation/src/main.c:142]: you've totally lost me

answer the question: why are you testing the ADDRESS of a pointer variable on the stack and not its contents? This is basic C coding concepts.

[the remark this answers is not in the document]
Comment (maksimmasalski) [reply tests/kernel/mem_protect/obj_validation/src/main.c:142]: I understand that I should use * to access data through the pointer which stores the address of the variable. Here I was wrong, I can't explain why exactly I wrote that line in the past.

File: tests/kernel/mem_protect/mem_protect/src/kobject.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (maksimmasalski) [reply tests/kernel/mem_protect/mem_protect/src/kobject.c]: @andrewboie I made changes. Could you take a look at test `void test_krnl_obj_static_alloc_build_time` ?

File: tests/kernel/mem_protect/obj_validation/src/main.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (maksimmasalski) [reply tests/kernel/mem_protect/obj_validation/src/main.c]: I made changes, I understand now that using & in that test was a mistake, because 
```
zassert_not_null(ptr, msg, ...)
Assert that ptr is not NULL.
```

**Comment by maksimmasalski** [comment]
> > I can't understand why Gitlint gives that error. I checked another tests use the same approach and define new typedefs.
> 
> That other test is wrong.
> 
> If the only reason you are defining a typedef is to save yourself from typing the keyword `struct` then don't do it.
> 
> Normally we only define typedefs for types that have to be truly opaque or are part of the assembly offset generation mechanism.

I found example here tests/kernel/queue/src/test_queue.h `typedef struct qdata`

**Comment by maksimmasalski** [comment]
> > I can't understand why Gitlint gives that error. I checked another tests use the same approach and define new typedefs.
> 
> That other test is wrong.
> 
> If the only reason you are defining a typedef is to save yourself from typing the keyword `struct` then don't do it.
> 
> Normally we only define typedefs for types that have to be truly opaque or are part of the assembly offset generation mechanism.

@andrewboie I resolved that too.

File: tests/kernel/mem_protect/userspace/src/main.c [hunk 3]
```diff
@@ -951,6 +959,49 @@ void test_tls_pointer(void)
 	}
 }
 
+void user_access_queue_private_data(void *p1, void *p2, void *p3)
+{
+	expect_fault = true;
+	expected_reason = K_ERR_KERNEL_OOPS;
+
+	/* try to access to private kernel data, will happen kernel oops */
+	k_queue_is_empty(&pqueue);
```
Comment (LeiW000) [inline tests/kernel/mem_protect/userspace/src/main.c:968]: My understanding is that kernel oops is expected behavior. But, how does sanitycheck can know if the kernel ops really happens to decide is the test case fails or not?

Comment (maksimmasalski) [reply tests/kernel/mem_protect/userspace/src/main.c:968]: Function supplied with the arguments for the function `void k_sys_fatal_error_handler`
```	
expect_fault = true;
expected_reason = K_ERR_KERNEL_OOPS;
```
If happened error has the same reason as the expected reason, then test passes.

**Review by LeiW000 — APPROVED** [review]

**Review by mengxianglinx — APPROVED** [review]

**Review by enjiamai — APPROVED** [review]

File: tests/kernel/mem_protect/obj_validation/src/main.c [hunk 4]
```diff
@@ -112,10 +121,109 @@ void test_generic_object(void)
...
+			 NULL, NULL, NULL,
+			 K_INHERIT_PERMS, 0, K_NO_WAIT);
+
+	k_thread_join(&user_thr0, K_FOREVER);
+
+	struct z_object *ko = z_object_find(test_dyn_mutex);
+
+	ko = z_object_find(test_dyn_mutex);
+
+	zassert_false(ko, NULL);
```
Comment (andrewboie) [inline tests/kernel/mem_protect/obj_validation/src/main.c:216]: please don't pass NULL to zassert_ functions for the description. Use a string that describes what happened and why it is not good.

File: tests/kernel/mem_protect/userspace/src/main.c [hunk 5]
```diff
@@ -951,6 +959,49 @@ void test_tls_pointer(void)
 	}
 }
 
+void user_access_queue_private_data(void *p1, void *p2, void *p3)
```
Comment (andrewboie) [inline tests/kernel/mem_protect/userspace/src/main.c:962]: can we put this in the k_queue tests and not in this test case?

I know there's some pipe code already here that needs to be moved as well

File: tests/kernel/mem_protect/userspace/src/main.c [hunk 6]
```diff
@@ -47,7 +50,12 @@ FOR_EACH(K_APPMEM_PARTITION_DEFINE, (;), part0, part1);
  */
 struct k_mem_domain dom0;
 struct k_mem_domain dom1;
+struct k_thread user_thread0;
+struct k_queue pqueue;
+
+static struct data_st test_data[LIST_LEN];
 
+K_APP_DMEM(part0) static volatile bool give_uthread_end_sem;
```
Comment (andrewboie) [inline tests/kernel/mem_protect/userspace/src/main.c:58]: why was this added? nothing seems to use it in this file unless I'm missing something

File: tests/kernel/mem_protect/obj_validation/src/main.c [hunk 7]
```diff
@@ -112,10 +121,109 @@ void test_generic_object(void)
...
+	struct k_thread *thread = _current;
+
+	intptr_t sem_addr = (intptr_t)&test_dyn_sem0;
+	intptr_t start_addr = (intptr_t)&thread->resource_pool;
+	size_t size_heap = CONFIG_HEAP_MEM_POOL_SIZE;
+
+	/* check that kernel object may be dynamically allocated,
+	 * drawing memory from the calling thread's memory pool
+	 */
+#if defined(CONFIG_X86)
```
Comment (andrewboie) [inline tests/kernel/mem_protect/obj_validation/src/main.c:153]: I don't understand why we have different cases here. resource pools are not implemented in an arch-specific manner.

File: tests/kernel/mem_protect/obj_validation/src/main.c [hunk 8]
```diff
@@ -112,10 +121,109 @@ void test_generic_object(void)
...
+	zassert_true(
+		(sem_addr > start_addr) && (sem_addr <= (start_addr + size_heap)),
+		"semaphore not in bound of thread's memory pool");
+#elif defined(CONFIG_ARM)
+	zassert_true(
+		(sem_addr > start_addr) && (sem_addr <= (start_addr + size_heap)),
+		"semaphore not in bound of thread's memory pool");
+#elif defined(CONFIG_ARC)
+	zassert_true(
```
Comment (andrewboie) [inline tests/kernel/mem_protect/obj_validation/src/main.c:162]: why does this need to be different from X86 and ARM?

File: tests/kernel/mem_protect/obj_validation/src/main.c [hunk 9]
```diff
@@ -112,10 +121,109 @@ void test_generic_object(void)
...
+		(sem_addr < start_addr) && (sem_addr > (start_addr - size_heap)),
+		"semaphore not in bound of thread's memory pool");
+#else
+#error "Not implemented for this architecture"
+	zassert_unreachable("Write to control register fault");
```
Comment (andrewboie) [inline tests/kernel/mem_protect/obj_validation/src/main.c:167]: Please don't copy-paste code. We are not writing to a control register

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
please don't pass NULL to the description argument to a zassert_ function in any test case, ever.

other problems noted inline

File: tests/kernel/mem_protect/obj_validation/src/main.c (hunk 4 above)
Comment (maksimmasalski) [reply tests/kernel/mem_protect/obj_validation/src/main.c:216]: Understood, I wrote code according to the existing zassert code that uses NULL and I thought it is acceptable.

File: tests/kernel/mem_protect/userspace/src/main.c (hunk 6 above)
Comment (maksimmasalski) [reply tests/kernel/mem_protect/userspace/src/main.c:58]: You are right, maybe during multiple rebasing I forgot to remove it.

Comment (maksimmasalski) [reply tests/kernel/mem_protect/userspace/src/main.c:58]: If you mean 
```
K_APP_DMEM(part0) static volatile bool give_uthread_end_sem;
K_APP_DMEM(part0) bool mem_access_check;
```
yes, it is just legacy and not used anywhere, I removed that.

That structures are used in test `test_access_kernel_obj_with_priv_data`:
```
struct k_thread user_thread0;
struct k_queue pqueue;

static struct data_st test_data[LIST_LEN];
```

File: tests/kernel/mem_protect/userspace/src/main.c (hunk 5 above)
Comment (maksimmasalski) [reply tests/kernel/mem_protect/userspace/src/main.c:955]: Do you mean move to tests/kernel/queue/?

Comment (maksimmasalski) [reply tests/kernel/mem_protect/userspace/src/main.c]: Done, in that PR added test to the kernel/queue directory.

**Comment by maksimmasalski** [comment]
I see that CI failure is not correct. Output from the failed boards show that no errors "PROJECT EXECUTION SUCCESSFU". Seems like CI bug.
mps2_an385 results and same for qemu_x86
```
<testcase classname="mps2_an385:tests/kernel/queue/kernel.queue" name="tests/kernel/queue/kernel.queue" time="2.257229">
<failure message="Failed" type="failure">*** Booting Zephyr OS build zephyr-v2.4.0-543-g0fde0c66acdc *** Running test suite queue_api =================================================================== START - test_queue_supv_to_user PASS - test_queue_supv_to_user =================================================================== START - test_queue_alloc_prepend_user PASS - test_queue_alloc_prepend_user =================================================================== START - test_queue_alloc_append_user PASS - test_queue_alloc_append_user =================================================================== START - test_auto_free PASS - test_auto_free =================================================================== START - test_queue_thread2thread PASS - test_queue_thread2thread =================================================================== START - test_queue_thread2isr PASS - test_queue_thread2isr =================================================================== START - test_queue_isr2thread PASS - test_queue_isr2thread =================================================================== START - test_queue_get_2threads PASS - test_queue_get_2threads =================================================================== START - test_queue_get_fail PASS - test_queue_get_fail =================================================================== START - test_queue_loop * Pass data by queue in loop 0 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 1 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 2 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 3 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 4 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 5 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 6 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 7 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 8 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 9 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 10 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 11 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 12 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 13 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 14 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 15 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 16 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 17 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 18 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 19 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 20 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 21 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 22 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 23 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 24 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 25 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 26 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 27 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 28 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 29 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 30 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get * Pass data by queue in loop 31 main queue append ---> isr queue find and remove isr queue get isr queue append ---> thread queue find and remove thread queue get thread queue append ---> main queue find and remove main queue get PASS - test_queue_loop =================================================================== START - test_queue_alloc PASS - test_queue_alloc =================================================================== START - test_queue_poll_race PASS - test_queue_poll_race =================================================================== START - test_multiple_queues PASS - test_multiple_queues =================================================================== START - test_access_kernel_obj_with_priv_data E: thread 0x20000390 (0) does not have permission on k_queue 0x20000ecc E: permission bitmap E: 00 00 |.. E: syscall z_vrfy_k_queue_is_empty failed check: access denied E: r0/a1: 0x00000000 r1/a2: 0x00000000 r2/a3: 0x00000000 E: r3/a4: 0x00000000 r12/ip: 0x00000000 r14/lr: 0x00000000 E: xpsr: 0x00000000 E: Faulting instruction address (r15/pc): 0x00000000 E: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0 E: Current thread: 0x20000390 (unknown) Caught system error -- reason 3 System error was expected PASS - test_access_kernel_obj_with_priv_data =================================================================== Test suite queue_api succeeded =================================================================== PROJECT EXECUTION SUCCESSFUL </failure>
</testcase>
```

**Comment by maksimmasalski** [comment]
@andrewboie I moved kernel/queue test to the separate PR #29188

File: tests/kernel/mem_protect/userspace/src/main.c (hunk 5 above)
Comment (maksimmasalski) [reply tests/kernel/mem_protect/userspace/src/main.c]: @andrewboie I moved kernel/queue test to the separate PR #29188

**Review by tyrantlizard77 — APPROVED** [review]
